### PR TITLE
Fix right-shifting deletions on AA sequence (fixes #498)

### LINKED ISSUE #498: Faulty right-shifting of deletions for protein annotations
## Summary

It seems that [`AminoAcidChangeNormalizer#normalizeDeletion`](https://github.com/charite/jannovar/blob/2472230f2428748777534e684c37f2ff5bf43f36/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/AminoAcidChangeNormalizer.java#L24) currently does not correctly right-shift protein sequences, introducing erroneous protein change HGVS annotations. Other kinds of variants (besides deletions) may be affected as well.

## Problem

The current code addresses the problems with deletions introducing frameshifts by avoiding right-shifting in those cases altogether, by checking [`change.getAlt().length() != 0`](https://github.com/charite/jannovar/blob/2472230f2428748777534e684c37f2ff5bf43f36/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/AminoAcidChangeNormalizer.java#L33).
This may be wrong if the frame-shifted sequence introduces the same amino acid(s) that would have followed in the wild type as well. It also does not address the following corner case:

- consider a deletion coinciding with the start of a codon, so that [`changeBeginPos.getFrameshift() == 0`](https://github.com/charite/jannovar/blob/6f6f8b4ce2f5ee27d607cdebac1dcebc0b2c9c07/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/builders/DeletionAnnotationBuilder.java#L133)
in the calling code (`DeletionAnnotationBuilder`), which means that `insAA` will be empty and thus normalization will be attempted by [`AminoAcidChangeNormalizer#normalizeDeletion`](https://github.com/charite/jannovar/blob/2472230f2428748777534e684c37f2ff5bf43f36/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/AminoAcidChangeNormalizer.java#L24)
- assume the length of the deletion is not divisible by three, so it will _itself_ introduce a frameshift, so that `varAASeq` (the AA sequence containing the variant) will differ significantly from from `wtAASeq` (the wildtype AA sequence) at the position of the deletion (and downstream)
- since `AminoAcidChangeNormalizer#normalizeDeletion` only operates on the _wildtype_ AA sequence (`wtAASeq`), this goes unnoticed, and in case the sequence of deleted amino acids starts with the same amino acid(s) as those _after_ the deletion _in the wild type_, the variant will be erroneously right-shifted on that wild-type sequence (the right-shifting currently does not consider `varAASeq` at all)

## Example (for corner case, see above)

Consider the following deletion (from a unit test in the PR attempting to fix this issue). 

```
Wildtype AA sequence:      H | T | G | E | K | P |fs F| E | C |
                     GGAAACAT|ACT|GGG|GAG|AAA|CCC| TTT|GAG|TGT|CCCAAATGTGGGAAGTGTTACTTTCGGAAGGAGA..
                     GGAAACAT|ACT|GGG|---|---|---|-TTG|AGT|GTC|CCAAATGT..
                                                 |  L | S | V |
```

Before [`AminoAcidChangeNormalizer#normalizeDeletion`](https://github.com/charite/jannovar/blob/2472230f2428748777534e684c37f2ff5bf43f36/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/AminoAcidChangeNormalizer.java#L24) gets called, `AminoAcidChange` will be `'EKPF'>''`. However, due to the above problem this will be _right-shifted_ to `KPFE>''` (as the first `E` in the wildtype sequence gets compared to the second `E` in the wildtype in the above example; afterwards `K` will be compared to `C`, which is wrong, so that `shift = 1`). 

This approach leads to a wrong position and a wrong AA displayed in the variant's protein change HGVS annotation. In this example, the protein change should be `p.(Glu316Leufs*25)` (in one-letter codes, `p.(E316Lfs*25)`) but is erroneously changed to `p.(Lys317Serfs*24)` (`p.(K317Sfs*24)`).

## Solution: more general approach for right-shifting

In contrast to the current right-shifting code for nucleotides, it seems a better approach here would be to simply _compare_ both wild-type and variant AA sequence, as they already have been computed by the `AminoAcidChangeNormalizer` anyhow. This approach has been implemented and tested in the accompanying PR (#499).

Now, right-shifting _also_ happens in case a deletion does not coincide with a codon start (i.e. `insAA` is non-empty, as checked by the previous code to avoid normalization). This is necessary because the resulting sequence in the variant AA sequence (`varAASeq`) could _still_ continue with the same amino acids as the wild type sequence would have, even for deletions that do not coincide with the beginning of a codon. For example:

```
Wildtype AA sequence:      H | T | G | E | K | P | F | E | C |
                     GGAAACAT|ACT|GGG|GAG|AAA|CCC|TTT|GAG|TGT|CCCAAATGTGGGAAGTGTTACTTTCGGAAGGAGA..
                     GGAAACAT|ACT|GG-|--G|AAA|CCC|TTT|AGT|GTC|CCAAATGT..
                                     | G | K | P |
```

This deletion would not have been normalized before, although the deletion results in the varant AA sequence continuing with the same AA (`G`) and thus the variant needs to be right-shifted by one.

## Related issues

- it seems #481 may be related to this issue
- if the above reasoning is correct, the code on which [`InsertionAnnotationBuilder` relies for protein sequence normalization](https://github.com/charite/jannovar/blob/2472230f2428748777534e684c37f2ff5bf43f36/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/AminoAcidChangeNormalizer.java#L88) may need to be changed correspondingly (happy to help here).

### PR DESCRIPTION
**Pull request by roland-ewald**
(see #498 for a summary and a nicer explanation with links to the affected source code)

## Problem

The previous code addressed the problems with deletions introducing frameshifts by avoiding right-shifting in those cases altogether, by checking `change.getAlt().length() != 0` in `AminoAcidChangeNormalizer#normalizeDeletion`. This may be wrong if the frame-shifted sequence introduces the same amino acid(s) that would have followed in the wild type as well.

It also does not address the following corner case:

- consider a deletion coinciding with the start of a codon, so that `changeBeginPos.getFrameshift() == 0` in the calling code
(`DeletionAnnotationBuilder`), which means that `insAA` will be empty and thus normalization will be attempted by
`AminoAcidChangeNormalizer#normalizeDeletion`

- assume the length of the deletion is not divisible by three, so it will itself introduce a frameshift, so that `varAASeq` (the AA sequence containing the variant) will differ significantly from from `wtAASeq` (the wildtype AA sequence) at the position of the deletion (and downstream)

- since `AminoAcidChangeNormalizer#normalizeDeletion` only operates on the wildtype AA sequence (`wtAASeq`), this goes unnoticed, and in case the sequence of deleted amino acids starts with the same amino acid(s) as those after the deletion in the wild type, the variant will be erroneously right-shifted on that wild-type sequence (the right-shifting currently did not consider `varAASeq` at all)

## Example

```
      H | T | G | E | K | P |fs F| E | C |
GGAAACAT|ACT|GGG|GAG|AAA|CCC| TTT|GAG|TGT|CCCAAATGTGGGAAGTGTTACTTTCGG...
GGAAACAT|ACT|GGG|---|---|---|-TTG|AGT|GTC|CCAAATGT..
                            |  L | S | V |
```

Before `AminoAcidChangeNormalizer#normalizeDeletion` gets called, `AminoAcidChange` will be `'EKPF'>''`. However, due to the above problem this will be right-shifted to `'KPFE'>''`, which leads to a wrong position and a wrong AA displayed in the variant's protein change HGVS annotation.

In this example, the protein change should be `p.(Glu316Leufs*25)` (in one-letter codes, `p.(E316Lfs*25)`) but is erroneously changed to `p.(Lys317Serfs*24)` (`p.(K317Sfs*24)`).

## Solution

In contrast to the right-shifting code for nucleotides, it seems a better approach here would be to simply compare both wild-type and variant AA sequence, as they already have been computed by the `AminoAcidChangeNormalizer` anyhow. So, `varAASeq` is now passed into `AminoAcidChangeNormalizer#normalizeDeletion` and compared to the wildtype AA sequence. This also allows us to apply right-shifting to deletions that do not coincide with the beginning of a codon.

## Tests

This adds both an 'integration' test with the above example to `DeletionAnnotationBuilderTest` and some unit tests to 
`AminoAcidChangeTest`. The latter has also been adjusted to make it clear that the tested operations should work on AA sequences, not nucleotide sequences (globally replaced 'A' by 'L'). 

Also note that at least one test (`testRealWorldCase_uc011mcs_()`) is failing with the changes introduced here. I'm not sure if this is a problem with the test or with my code, but from looking at the difference between expected and actual outcome, it seems the current test data is wrong (but I wanted to confirm the whole issue first, before digging in deeper).

## to do

_(not sure whether I should be doing this, but these points should be resolved before a merge)_

- [ ] confirm this is an issue and the approach taken here is reasonable
- [ ] adjust test data so that tests pass
- [ ] adjust (and test) other areas with a similar issue, in particular check `InsertionAnnotationBuilder`

### CODE REVIEW

File: jannovar-core/src/test/java/de/charite/compbio/jannovar/reference/AminoAcidChangeTest.java [hunk 1]
```diff
@@ -13,42 +13,56 @@
 
 	@Before
 	public void setUp() throws Exception {
-		this.ref = "AAACCCAAACCC";
-		this.ref2 = "CATCATCTTCA";
+		this.ref = "LLLCCCLLLCCC";
```
Comment (roland-ewald) [inline jannovar-core/src/test/java/de/charite/compbio/jannovar/reference/AminoAcidChangeTest.java:16]: This is just 'cosmetics', just making them look 'more' like AA sequences than like a nucleotide sequence (of course the previous sequence was also a valid AA sequence)

**Comment by roland-ewald** [comment]
Just as a quick update, I have checked `InsertionAnnotationBuilder` and `BlockSubstitutionAnnotationBuilder` and could reproduce the issue there as well. With a slightly generalized method for right-shifting the AA sequence given `wtAASeq` and `varAASeq`, the same shifting method can be called in all three cases to fix the issue (AFAICT). Let me know if/when to update the PR accordingly.

**Comment by holtgrewe** [comment]
HI @roland-ewald, I'm sorry for the long delay in answering. I'm now committed to push through this pull request, however, and would be very grateful if you could pick up your great work here.

> Let me know if/when to update the PR accordingly.

If this offer still stands then I would be very happy if you could update the PR.

**Comment by roland-ewald** [comment]
Hi @holtgrewe , no worries :+1: 
I'm a little short on time right now but will try to find my old notes this weekend, and then make some progress. So, this issue and the suggested approach to fix it (ie., using a single method for all three builders) makes sense to you?

**Comment by holtgrewe** [comment]
Thanks, sounds like an excellent plan.

File: jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/AminoAcidChangeNormalizer.java [hunk 2]
```diff
@@ -21,46 +24,44 @@ public static AminoAcidChange truncateAltAfterStopCodon(AminoAcidChange change)
...
 	 * <p>
 	 * Truncating of the prefixes is given higher priority to conform with the HGVS notation (you have to call
-	 * {@link #shiftInsertion}) afterwards.
+	 * {@link #shiftSynonymousChange}) afterwards.
 	 *
 	 * @param aaChange {@link AminoAcidChange} to truncate
 	 * @return updated {@link AminoAcidChange}
 	 */
-	public static AminoAcidChange truncateBothSides(AminoAcidChange aaChange) {
-		// TODO(holtgrem): Test me!
```
Comment (roland-ewald) [inline jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/AminoAcidChangeNormalizer.java:61]: I've removed this TODO since this is now a 'helper method' for `shiftSynonymousChange` and thus tested with it; I made sure to add additional test data that covers such cases to `AminoAcidChangeTest`.

File: jannovar-core/src/test/java/de/charite/compbio/jannovar/annotation/builders/DeletionAnnotationBuilderTest.java [hunk 3]
```diff
@@ -1221,7 +1221,7 @@ public void testRealWorldCase_uc011mcs_() throws InvalidGenomeVariant {
 		Assert.assertEquals(infoForward.getAccession(), annotation1.getTranscript().getAccession());
 		Assert.assertEquals(12, annotation1.getAnnoLoc().getRank());
 		Assert.assertEquals("1068_1071del", annotation1.getCDSNTChange().toHGVSString());
-		Assert.assertEquals("(Glu358del)", annotation1.getProteinChange().toHGVSString(AminoAcidCode.THREE_LETTER));
+		Assert.assertEquals("(Glu357Lysfs*?)", annotation1.getProteinChange().toHGVSString(AminoAcidCode.THREE_LETTER));
```
Comment (roland-ewald) [inline jannovar-core/src/test/java/de/charite/compbio/jannovar/annotation/builders/DeletionAnnotationBuilderTest.java:1224]: I wasn't able to verify _this specific_ variant, just changed this to the output now generated by the fixed method. Maybe the best strategy here would be to use a 'known valid' protein change description as ground truth (e.g. from ClinVar) and add another test case.

File: jannovar-core/src/test/java/de/charite/compbio/jannovar/reference/AminoAcidChangeTest.java [hunk 4]
```diff
@@ -20,49 +20,77 @@ public void setUp() throws Exception {
...
-	public void testShiftFrameshiftInsAA() {
+	public void testShiftDeletionFrameshiftInsAA() {
 		AminoAcidChange origChange = new AminoAcidChange(3, "CQY", "C");
-		AminoAcidChange modChange = AminoAcidChangeNormalizer.normalizeDeletion("LLLCQYCLLL", "LLLCQYVA", origChange);
-		Assert.assertEquals(new AminoAcidChange(6, "CLL", "V"), modChange);
+		AminoAcidChange modChange = AminoAcidChangeNormalizer.shiftSynonymousChange(origChange, "LLLCQYCLLL", "LLLCQYVA");
+		Assert.assertEquals(new AminoAcidChange(6, "CL", ""), modChange);
+	}
+
+	@Test
```
Comment (roland-ewald) [inline jannovar-core/src/test/java/de/charite/compbio/jannovar/reference/AminoAcidChangeTest.java:55]: Added these tests as `AminoAcidChangeNormalizer.shiftSynonymousChange` is now also used for insertions and block substitutions.

**Review by roland-ewald — COMMENTED** [review]
@holtgrewe I've changed the other affected parts of the code base now, refactored the new code accordingly, and gave the method another name that should be more suitable for the new callers (`Insertion/BlockSubstitutionAnnotationBuilder`): `shiftSynonymousChange`. Feel free to change! 

Since `AminoAcidChangeNormalizer.truncateBothSides` is always called just before this method gets called, I made it private and call it directly from `shiftSynonymousChange`. This simplifies the code in the annotation builders a little, avoids a potential error source (forgetting to truncate), and also allows the tests in `AminoAcidChangeTest` to cover the method as well. Hope that's ok, otherwise feel free to roll this back.

Hope this helps.

**Review by holtgrewe — COMMENTED** [review]
All of the changes look good. I will add to the HISTORY, rebase to current `master`, and then merge. I don't know whether I will be able to push to your downstream branch for this so it might look like I'm just closing this (but I am not).

The squash should still keep your contribution as co-authored-with.

**Review by holtgrewe — APPROVED** [review]

**Comment by roland-ewald** [comment]
@holtgrewe no worries; thanks for taking the time to review and merge this! :+1:
